Clarify that digests don't have to be cryptographic ones. (#338)
in-toto/attestation: spec/v1/digest_set.md

Reviewer: Could you revert the unrelated changes to formatting? Looks like you used a tool that formatted the entire file instead of just modified lines.

Author: Done. At some point I'd like to hear about your workflow here.  I'm just podging things together. :/

# DigestSet field type specification

> Version: v1.1

Author: Added

# DigestSet field type specification

Version: v1.1

> Set of one or more immutable digests for a single software artifact or metadata
> object.

Reviewer: nit: a digest is always immutable since it's just a string. I think "immutable reference" is more accurate. Also I think it may be worth calling out that the overwhelming majority use case is cryptographic digest, to avoid confusion.

```suggestion
Set of one or more cryptographic digests, or other immutable references, for a
single software artifact or metadata object.
```

(Also ok to just say "immutable references" - your choice)

Author: Done

## Schema

```json
{
  "<ALGORITHM_1>": "<VALUE>",
  "<ALGORITHM_2>": "<VALUE>",
  ...
}
```

## Fields

A DigestSet is represented as a *JSON object* mapping algorithm name to a string
encoding of the digest using that algorithm. The named standard algorithms below
use lowercase hex encoding. Usually there is just a single key/value pair, but
multiple entries MAY be used for algorithm agility.

> Users SHOULD use a *cryptographic* digest, but MAY use another identifier if the
> underlying implementation ensures immutability via other means.

Reviewer: You might want to spell this out a bit more, since it's a bit nuanced. Maybe:

```suggestion
Each entry in a DigestSet MUST be an immutable reference to an artifact. It is
STRONGLY RECOMMENDED to use a commonly accepted, cryptographically secure digest
algorithm to achieve this immutability. See [Use cases for non-cryptographic,
immutable, digests](...) for further guidance.
```

Author: Done.

single key/value pair, but multiple entries MAY be used for algorithm
agility.

Each entry in a DigestSet MUST be an immutable reference to an artifact. It is
STRONGLY RECOMMENDED to use a commonly accepted, cryptographically secure digest
algorithm to achieve this immutability. See [Use cases for non-cryptographic,
immutable, digests](#use-cases-for-non-cryptographic-immutable-digests) for
further guidance.

Users SHOULD use a _cryptographic_ digest, but MAY use another identifier

> if the underlying implementation ensures immutability via other means.

Reviewer 2: []()

### Guidelines

It is RECOMMENDED to use at least `sha256` for compatibility between producers
and consumers, unless a different hash algorithm is more conventional (e.g.
`gitCommit` for git).

Consumers MUST only accept algorithms that they consider secure and MUST ignore
unrecognized or unaccepted algorithms. For example, most applications SHOULD NOT
accept "md5" because it lacks collision resistance.

Two DigestSets SHOULD be considered matching if ANY acceptable field matches.

New algorithms MUST document how the value is encoded, e.g. URL-safe base64,
lowercase hex, etc...

### Use cases for non-cryptographic, immutable, digests

> Sometimes users have a need to refer to something by some other immutable
> identifier. Either because the content can't be hashed traditionally, because
> it's impractical to hash traditionally, or because they interact with the
> content through an interface that doesn't expose them to the entirety of the
> content.
> 
> In these situations users may wish to use other identifiers in a DigestSet.
> Those users should be careful to understand the trust that they're placing in
> the identifier to be sure that it meets their needs.

Reviewer: Some editing suggestions to improve readability. Feel free to accept partially or to edit further.

```suggestion
While cryptographic digests are the strongly recommended immutable identifier,
users might have need to refer to an artifact by some other means. For example,
it might be technically infeasible to compute a digest over the content, or
because the user might interact with the content through an interface that
doesn't expose them to the entirety of the content.

In these situations, users MAY use a non-cryptographic identifier in a DigestSet so
long as the risk of the object being mutated is acceptable for the application.
```

Author: Done
